extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetryableS3OutputStream.java

```diff
@@ -103,22 +107,32 @@ public class RetryableS3OutputStream extends OutputStream
   private boolean error;
   private boolean closed;
 
-  public RetryableS3OutputStream(
-      S3OutputConfig config,
-      ServerSideEncryptingAmazonS3 s3,
-      String s3Key
-  ) throws IOException
-  {
+  /**
+   * An atomic counter to store number of files pending to be uploaded for the particular uploadId.
+   */
+  private final AtomicInteger pendingFiles = new AtomicInteger(0);
 
-    this(config, s3, s3Key, true);
-  }
+  /**
+   * A lock used for notifying the main thread about the completion of s3.uploadPart() for all chunks
+   * and hence starting the s3.completeMultipartUpload() for the uploadId.
+   */
+  private final Object fileLock = new Object();
+
+  /**
+   * Semaphore to restrict the maximum number of simultaneous chunks on disk.
+   */
+  private static final int MAX_CONCURRENT_CHUNKS = 10;
+  private static final Semaphore SEMAPHORE = new Semaphore(MAX_CONCURRENT_CHUNKS);
 
-  @VisibleForTesting
-  protected RetryableS3OutputStream(
+  /**
+   * Threadpool used for uploading the chunks asynchronously.
+   */
+  private static final ExecutorService UPLOAD_EXECUTOR = Execs.multiThreaded(10, "UploadThreadPool-%d");
```

Comment:
This executor should not be static. If the intention is to share the executor, it should be passed in as an argument to the constructor. Also, there should be a well defined lifecycle of the executor, i.e. when it is started and stopped.

Reply:
> This executor should not be static. If the intention is to share the executor, it should be passed in as an argument to the constructor.

But then we'd have to inject it and pass it through a bunch of files: S3StorageConnectorModule, S3ExportStorageProvider, S3StorageConnectorProvider, StorageConnector, RetryableS3OutputStream.

Seems unnecessary to me? For my understanding, what benefit does it provide over the current approach?

> Also, there should be a well defined lifecycle of the executor, i.e. when it is started and stopped.

Where would you suggest to shutdown() the executor?

Reply:
> Seems unnecessary to me? For my understanding, what benefit does it provide over the current approach?

It is not unnecessary, it is the right approach 🙂 . `static` should be used for constants, static variables and static classes. It doesn't make sense to make executors and other resources as static for the sole purpose of sharing them across instances. It makes the code difficult to follow and interferes with the lifecycle of such resources. You never know when to clean up such a resource as you don't know if an instance still exists in the JVM.

An executor is a resource and if it is meant to be shared, it must be explicitly passed into the constructor as a dependency.

Reply:
> Where would you suggest to shutdown() the executor?

If the executor is meant to be used throughout the lifecycle of the underlying service, shutting it down on service shutdown would make sense. All Druid services are injected with a `ScheduledExecutorFactory` which can be used for creating an executor which is tied to the lifecycle of the service.

Reply:
Thanks for the explanation, makes sense!

> All Druid services are injected with a ScheduledExecutorFactory which can be used for creating an executor

I don't see any occurrence of this in the `druid-s3-extensions`? Can you suggest how to use this?

Reply:
I tried adding the following in `S3StorageConnectorModule`:
```java
@Provides
  @LazySingleton
  public ScheduledExecutorFactory getScheduledExecutorFactory(Lifecycle lifecycle)
  {
    return ScheduledExecutors.createFactory(lifecycle);
  }
```
...and then `@Inject ScheduledExecutorFactory scheduledExecutorFactory;` in `S3StorageConnectorProvider`, but it's null.

Can you advise on how/where to inject it to be available in the `druid-s3-extensions`? Thanks!

Reply:
In `S3StorageConnectorProvider`, you would need to use `@JacksonInject` instead of plain `@Inject`.
Also, you probably don't need to add the provider method `getScheduledExecutorFactory` as it is already being provided in `ServerModule`.

Reply:
Have injected ExecutorService created via ScheduledExecutorFactory. Hope this works.